pkg/vm/builtins/modulo.go

```diff
@@ -464,78 +479,122 @@ func (m *ModBuiltin) fillValue(mem *memory.Memory, inputs ModBuiltinInputs, inde
-	}
-	mulModBuiltinSegment, ok := mem.FindSegmentWithBuiltin("MulMod")
-	if !ok {
-		return fmt.Errorf("MulMod builtin segment doesn't exist")
-	}
-	addModBuiltinRunner, ok := addModBuiltinSegment.BuiltinRunner.(*ModBuiltin)
-	if !ok {
-		return fmt.Errorf("addModBuiltinRunner is not a ModBuiltin")
-	}
-	mulModBuiltinRunner, ok := mulModBuiltinSegment.BuiltinRunner.(*ModBuiltin)
-	if !ok {
-		return fmt.Errorf("mulModBuiltinRunner is not a ModBuiltin")
-	}
+	var addModBuiltinRunner *ModBuiltin
+	var mulModBuiltinRunner *ModBuiltin
 
-	if addModBuiltinRunner.wordBitLen != mulModBuiltinRunner.wordBitLen {
-		return fmt.Errorf("AddMod and MulMod wordBitLen mismatch")
+	if nAddMods != 0 {
+		addModBuiltinSegment, ok := mem.FindSegmentWithBuiltin("AddMod")
+		if !ok {
+			return fmt.Errorf("AddMod builtin segment doesn't exist")
+		}
+		addModBuiltinRunner, ok = addModBuiltinSegment.BuiltinRunner.(*ModBuiltin)
+		if !ok {
+			return fmt.Errorf("addModBuiltinRunner is not a ModBuiltin")
+		}
+	} else {
+		addModBuiltinRunner = nil
 	}
 
-	addModBuiltinInputs, err := addModBuiltinRunner.readInputs(mem, addModBuiltinAddr, true)
-	if err != nil {
-		return err
-	}
-	if err := addModBuiltinRunner.fillInputs(mem, addModBuiltinAddr, addModBuiltinInputs); err != nil {
-		return err
-	}
-	if err := addModBuiltinRunner.fillOffsets(mem, addModBuiltinInputs.offsetsPtr, nAddModsIndex, addModBuiltinInputs.n-nAddModsIndex); err != nil {
-		return err
+	if nMulMods != 0 {
+		mulModBuiltinSegment, ok := mem.FindSegmentWithBuiltin("MulMod")
+		if !ok {
+			return fmt.Errorf("MulMod builtin segment doesn't exist")
+		}
+		mulModBuiltinRunner, ok = mulModBuiltinSegment.BuiltinRunner.(*ModBuiltin)
+		if !ok {
+			return fmt.Errorf("mulModBuiltinRunner is not a ModBuiltin")
+		}
+	} else {
+		mulModBuiltinRunner = nil
 	}
 
-	mulModBuiltinInputs, err := mulModBuiltinRunner.readInputs(mem, mulModBuiltinAddr, true)
-	if err != nil {
-		return err
-	}
-	if err := mulModBuiltinRunner.fillInputs(mem, mulModBuiltinAddr, mulModBuiltinInputs); err != nil {
-		return err
-	}
-	if err := mulModBuiltinRunner.fillOffsets(mem, mulModBuiltinInputs.offsetsPtr, nMulModsIndex, mulModBuiltinInputs.n-nMulModsIndex); err != nil {
-		return err
-	}
+	var addModBuiltinInputs, mulModBuiltinInputs ModBuiltinInputs
+	var err error
 
-	addModIndex, mulModIndex := uint64(0), uint64(0)
-	for addModIndex < nAddModsIndex {
-		ok, err := addModBuiltinRunner.fillValue(mem, addModBuiltinInputs, int(addModIndex), Add)
+	if addModBuiltinRunner != nil {
+		addModBuiltinInputs, err = addModBuiltinRunner.readInputs(mem, addModInputAddress, true)
 		if err != nil {
 			return err
 		}
-		if ok {
-			addModIndex++
+		if err := addModBuiltinRunner.fillInputs(mem, addModInputAddress, addModBuiltinInputs); err != nil {
+			return err
+		}
+		if err := addModBuiltinRunner.fillOffsets(mem, addModBuiltinInputs.offsetsPtr, nAddMods, addModBuiltinInputs.n-nAddMods); err != nil {
+			return err
```

Comment:
Couldn't you add this logic in the first if? I mean, where you know that AddModBuiltinRunner is not nil. This is completely independent from the mulMod logic so I don't see why not

Reply:
How do you mean ?